sw/device/lib/testing/test_framework/ottf_console.c

```diff
@@ -419,12 +425,65 @@ size_t ottf_console_spi_device_read(size_t buf_size, uint8_t *const buf) {
   return received_data_len;
 }
 
-status_t ottf_console_putbuf(void *io, const char *buf, size_t len) {
+static status_t ottf_spi_console_flushbuf(void *io) {
+  size_t written_len = 0;
+  if (spi_buf_end > 0) {
+    written_len = sink(io, spi_buf, spi_buf_end);
+    if (spi_buf_end != written_len) {
+      return DATA_LOSS((int32_t)(spi_buf_end - written_len));
+    }
+    spi_buf_end = 0;
+  }
+  return OK_STATUS((int32_t)written_len);
+}
+
+static status_t ottf_spi_console_putbuf(void *io, const char *buf, size_t len) {
+  if (len > kSpiDeviceMaxFramePayloadSizeBytes) {
+    // Flush and skip the staging buffer if the payload is already the max size.
+    TRY(ottf_spi_console_flushbuf(io));
+    size_t written_len = sink(io, buf, len);
+    if (len != written_len) {
+      return DATA_LOSS((int32_t)(len - written_len));
+    }
+  } else if ((spi_buf_end + len - 1) < kSpiDeviceMaxFramePayloadSizeBytes) {
```

Comment:
Would this condition be clearer as `spi_buf_end + len <= kSpiDeviceMaxFramePayloadSizeBytes`?

Rationale for why it's clearer: these are all `size_t` items and we're computing lengths.  Less than or equal means the length of `buf` will fit.

Aside: my personal style would be to use `sizeof(spi_buf)` here because we're using `memcpy` (or any other byte-length accepting function).  The constant is perfectly fine; just pointing out an alternative.